tasks_week1_Gal_Malka/src/Vector.cpp

```diff
@@ -0,0 +1,355 @@
+*/
+int& Vector::operator[] (size_t index) {
+
+	if (index < 0 || index > _size - 1) {
+
+		throw std::out_of_range("Array index out of bounds.");
+	}
+	return *(_data + index);
+}
+
+/*
+* const operator[] - for random access to vector's values without changing any of the data.
+*/
+const int& Vector::operator[] (size_t index) const {
+
+	if (index < 0 || index > _size - 1) {
+
+		throw std::out_of_range("Array index out of bounds.");
+	}
+
+	return *(_data + index);
+}
+
+/*
+* operator= - copy the argument vector data into the vector and deallocate previous data.
+*/
+Vector& Vector::operator= (const Vector& other) {
+
+	if (this != &other) {
+
+		_size = other._size;
+		_capacity = other._capacity;
+		if(_data)
+			delete[] _data;
+		_data = new int[_capacity];
+		copy_data(other._data, &_data, other._size);
+	}
+
+	return *this;
+}
+
+/*
+* operator= - moves the argument vector data control into the vector and deallocate previous data.
+*/
+Vector& Vector::operator= (Vector&& other)noexcept {
+
+	if (this != &other) {
+
+		delete[] _data;
+		_size = other._size;
+		_capacity = other._capacity;
+		_data = other._data;
+		other._data = nullptr;
+	}
+
+	return *this;
+};
+
+/*
+* operator= - copy the initializer list data into the vector and deallocate previous data.
+*/
+Vector& Vector::operator= (const std::initializer_list<int> initializer) {
+	_size = initializer.size();
+	_capacity = _size;
+	delete[] _data;
+	_data = new int[_capacity];
+	copy_data(initializer.begin(), &_data, _size);
+	return *this;
+};
+
+/*
+* swap - swaps the data between two vectors.
+*/
+void Vector::swap(Vector& other) {
+
+	//swap capacity
+	other._capacity += _capacity;
+	_capacity = other._capacity - _capacity ;
+	other._capacity -= _capacity;
+
+	//swap size
+	other._size += _size;
+	_size = other._size - _size;
+	other._size -= _size;
+
+	//swap data
+	int* temp = other._data;
+	other._data = _data;
+	_data = temp;
+
+	temp = nullptr;
+}
+
+/*
+* push_back - append a value to the end of the vector. if the vector's capacity is not enaugh,
+* than it will allocate more memory.
+*/
+void Vector::push_back(const int& value) {
+
+	if (_size == _capacity) {
+
+		if (_capacity == 0) {
+
+			_capacity = 2;
+			_data = new int[_capacity];
+		}
+
+		else {
+			//allocate more memory and copy data.
+			_capacity = _capacity < 128 ? _capacity * 2 : _capacity * 1.5;
+			int* newData = new int[_capacity];
+			copy_data(_data, &newData, _size);
+			delete[] _data;
+			_data = newData;
+			newData = nullptr;
+		}
+	}
+
+	*(_data +_size) = value;
+	_size++;
+
+}	
+
+/*
+* clear - destroy all data
+*/
+void Vector::clear() noexcept {
+
+	_size = 0;
+	_capacity = 0;
+	if (_data)
+		delete[] _data;
+	_data = nullptr;
+}
+
+/*
+* resize - change the size of the vector to newSize
+*/
+void Vector::resize(size_t newSize) {
+
+	if (newSize > _capacity) {
+		//allocate more memory and copy data.
+		_capacity = newSize < 128 ? newSize * 2 : newSize * 1.5;

```

Comment:
This calculation is duplicated (saw it also in `push_back`). Reducing code duplication is essential for writing maintainable code. Think how you can remove this duplication